### PR TITLE
Remove a redundant heuristic

### LINKED ISSUE #5529: `cargo update -p SPEC` can silently downgrade dependencies of other packages
This is kinda convoluted, but it bit me when I wasn't paying attention when updating rust's lock file.  Here's a relatively small repro.  Create a workspace with the following members and dependencies:

```
a → bitflags="0.9"
b → bitflags=">= 0.9, < 2.0"
c → bitflags="1.0"
```

`cargo update` will generate a lockfile that looks good (`a`→0.9, `b`→1.0, `c`→1.0).

`cargo update -p c` will *downgrade* `b`'s copy of bitflags to 0.9!  I'd expect it to leave it as-is.

The real-world example of this is in rust where these chains exist:
```
rustbook → ... → html5ever → .. → rand 0.3
rls → rayon → rayon-core → rand >=0.3, < 0.5
cargo → ... → rand 0.4
```
Updating cargo caused rayon's version of rand to downgrade.  I didn't use `rand` in my example above because version 0.3 is a little odd in that it also depends on rand 0.4, but doesn't seem to be important and just complicates the example.

### LINKED ISSUE #5702: dependency downgrade on cargo update
```
cargo 1.26.0 (0e7c5a931 2018-04-06)
```

-----

I run `cargo update -p snappy` on [parity](https://github.com/paritytech/parity) project  - rev `bca100cdb047906e73d6733f43df73c7603897dd`.

cargo output was ok:

```
    Updating git repository `https://github.com/paritytech/rust-snappy`
    Updating git repository `https://github.com/paritytech/ring`
    Updating registry `https://github.com/rust-lang/crates.io-index`
      Adding cmake v0.1.31
    Updating snappy v0.1.0 (https://github.com/paritytech/rust-snappy#40ac9a0d) -> #798408ff
    Updating snappy-sys v0.1.0 (https://github.com/paritytech/rust-snappy#40ac9a0d) -> #798408ff
```

but to my surprise, one of my dependencies has been downgraded with no apparent reason

```diff
diff --git a/Cargo.lock b/Cargo.lock
index c791ae4..d69a4c7 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -246,6 +246,14 @@ dependencies = [
 ]
 
 [[package]]
+name = "cmake"
+version = "0.1.31"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "cc 1.0.17 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
 name = "common-types"
 version = "0.1.0"
 dependencies = [
@@ -1969,7 +1977,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
  "num-integer 0.1.36 (registry+https://github.com/rust-lang/crates.io-index)",
  "num-traits 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
- "rand 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)",
+ "rand 0.3.20 (registry+https://github.com/rust-lang/crates.io-index)",
  "rustc-serialize 0.3.24 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
 
@@ -2765,7 +2773,7 @@ dependencies = [
  "lazy_static 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)",
  "libc 0.2.36 (registry+https://github.com/rust-lang/crates.io-index)",
  "num_cpus 1.8.0 (registry+https://github.com/rust-lang/crates.io-index)",
- "rand 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)",
+ "rand 0.3.20 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
 
 [[package]]
@@ -3118,7 +3126,7 @@ dependencies = [
 [[package]]
 name = "snappy"
 version = "0.1.0"
-source = "git+https://github.com/paritytech/rust-snappy#40ac9a0d9fd613e7f38df800a11a589b7296da73"
+source = "git+https://github.com/paritytech/rust-snappy#798408ffef8f86dd51481673aca10f5348d7491b"
 dependencies = [
  "libc 0.2.36 (registry+https://github.com/rust-lang/crates.io-index)",
  "snappy-sys 0.1.0 (git+https://github.com/paritytech/rust-snappy)",
@@ -3127,9 +3135,9 @@ dependencies = [
 [[package]]
 name = "snappy-sys"
 version = "0.1.0"
-source = "git+https://github.com/paritytech/rust-snappy#40ac9a0d9fd613e7f38df800a11a589b7296da73"
+source = "git+https://github.com/paritytech/rust-snappy#798408ffef8f86dd51481673aca10f5348d7491b"
 dependencies = [
- "cc 1.0.17 (registry+https://github.com/rust-lang/crates.io-index)",
+ "cmake 0.1.31 (registry+https://github.com/rust-lang/crates.io-index)",
  "libc 0.2.36 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
 
@@ -3942,6 +3950,7 @@ dependencies = [
 "checksum chrono 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)" = "1cce36c92cb605414e9b824f866f5babe0a0368e39ea07393b9b63cf3844c0e6"
 "checksum cid 0.2.3 (registry+https://github.com/rust-lang/crates.io-index)" = "d85ee025368e69063c420cbb2ed9f852cb03a5e69b73be021e65726ce03585b6"
 "checksum clap 2.29.1 (registry+https://github.com/rust-lang/crates.io-index)" = "8f4a2b3bb7ef3c672d7c13d15613211d5a6976b6892c598b0fcb5d40765f19c2"
+"checksum cmake 0.1.31 (registry+https://github.com/rust-lang/crates.io-index)" = "95470235c31c726d72bf2e1f421adc1e65b9d561bf5529612cbe1a72da1467b3"
 "checksum conv 0.3.3 (registry+https://github.com/rust-lang/crates.io-index)" = "78ff10625fd0ac447827aa30ea8b861fead473bb60aeb73af6c1c58caf0d1299"
 "checksum crc 1.7.0 (registry+https://github.com/rust-lang/crates.io-index)" = "bd5d02c0aac6bd68393ed69e00bbc2457f3e89075c6349db7189618dc4ddc1d7"
 "checksum crossbeam 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)" = "24ce9782d4d5c53674646a6a4c1863a21a8fc0cb649b3c94dfc16e45071dea19"
```

-----

`rand` crate has been downgraded for `num-bigint` and `rayon-core` which require `rand >= 0.3, < 0.5`

### PR DESCRIPTION
**Pull request by Eh2406**
This adds a test for #5529, then removes the heuristic that caused that odd behavior. 
I believe this heuristic is redundant with what the resolver already does. 

If you believe this heuristic is needed please suggest a test case and I will add it to this PR. 

This tests and closes #5529 and should close #5702.

### CODE REVIEW

**Comment by Eh2406** [comment]
Looks like some of these test cases are already in. let me investegate.

**Comment by alexcrichton** [comment]
Ah yeah I believe the failing tests are why this logic was here, but it should definitely be fine to move around the logic to keep the tests passing!

**Comment by Eh2406** [comment]
As usual when code looks redundant there's a corner case that has not been considered. In this case the fact that was not in my mental model was that querying some dependencies can have observable side effects. Specifically, git dependencies will be pulled too the latest commit.

**Comment by Eh2406** [comment]
Ok so there is an externality for querying a Crates.io dependency. It triggers a registry update #2895. (Although If I wanted to do an update but not update the registry, I'd think `--offline` is a better suggestion now that we have it.)

I seem to be out of ideas.

**Comment by alexcrichton** [comment]
Hm in general I've tried to minimize in Cargo the number of locations where the type of source is specialized, ideally all of them should be total black boxes to the resolver and Cargo. (that doesn't actually manifest currently but it's somewhat close!)

I'm not 100% sure that https://github.com/rust-lang/cargo/issues/5529 is a bug per se in the sense that both crate graphs are "correct" and Cargo doesn't have much ability to pick between the two to say which is "more correct". It may be the case though that we could add a new heuristic for sorting dependencies on whether the dependency will cause a slow update? (something like fetching the index or updating a git repo)

One thing we could also try is adding dummy entries from the lockfile into the "can be locked to" set in the registry. That way we may end up throwing them away if nothing uses it, but it's at least there as a candidate to be used?

**Comment by Eh2406** [comment]
> I'm not 100% sure that #5529 is a bug per se

I'd think that a quickchek like invariant is that if a lockfile was newly generated and nun of the inputs have changed then all `cargo update ...` commands will not change the lock file. But given that the heuristics in the resolver are order dependant, I am not sure that this invariant is true.

> One thing we could also try is adding dummy entries from the lockfile into the "can be locked to" set in the registry. That way we may end up throwing them away if nothing uses it, but it's at least there as a candidate to be used?

Rufly this heuristic should be looking at `union(things locked, things preferred by conservative update)` instead of just `things locked`?

**Comment by alexcrichton** [comment]
@Eh2406 I believe so yeah! 

Awhile back I made some tweaks which were targeted at fixing errors during an update which otherwise didn't need to happen, and along the way what I ended up doing was to basically sort version candidates to the front if they were already in the lock file (preferring those) over other candidates. The upside of this is that most of the time the candidates in the lock file always work, but in the rare case they cause conflicts or something like that we still have the other candidates to try. 

I think is similar in the sense that we should first try what was already in the lock file (in theory) or whatever candidates are available in the lock file. If that all fails though we should fall back to normal version resolution and update the registry and such. In theory this is sort of like an iterator where each call to `next()` may return a long time, and it's sorted by lockfile-first which don't need to update the registry.

I'm not sure how much of that's easy to implement though :(

**Comment by Eh2406** [comment]
I very agree! I really like your `move things in lockfile to the front` heuristic! The thrust of this PR is that we should rely on it more, and reduce the use of `="lockfile ver"` heuristics.

Naively removing this heuristic did not work, as it indirectly signals some externalities not to trigger. So what we want is for the resolver to backtrack from querying with out externalities to querying with externalities. I have been thinking on that and have not figured out how to teach the resolver to do that. :-( 

But I do think we can add a new layer of backtracking, If we add an argument to querying for whether to trigger externalities then we can do a conservative update by calling the resolver with that argument set to no externalities if it succeeds we are good, if not we can rerun the resolver argument set to use externalities. 

I think this will work but have not tried it yet. I don't know when I will so I left this as notes for futcher me.

**Comment by alexcrichton** [comment]
> So what we want is for the resolver to backtrack from querying with out externalities to querying with externalities

I think this is exactly the problem, yeah!

I'm down for a solution to this so long as it keeps test passing, so feel free to go wild!

**Comment by alexcrichton** [comment]
I'm gonna close this for now to help clear out the queue, but I'm all for merging w/ the last point figured out!
